iscsi/get.go

```diff
@@ -0,0 +1,314 @@
+// Copyright 2017 The Prometheus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iscsi
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+// GetStats is the main iscsi status information func
+// building the path and prepare info for enable iscsi
+func GetStats(iqnPath string) (*Stats, error) {
+	var istats Stats
+
+	istats.Name = filepath.Base(iqnPath)
+	istats.RootPath = filepath.Dir(iqnPath)
+
+	matches, err := filepath.Glob(filepath.Join(iqnPath, "tpgt*"))
+	if err != nil {
+		return nil, fmt.Errorf("iscsi: GetStats: get TPGT path error %v", err)
+	}
+	istats.Tpgt = make([]TPGT, len(matches))
+
+	for ii, tpgtPath := range matches {
+		istats.Tpgt[ii].Name = filepath.Base(tpgtPath)
+		istats.Tpgt[ii].TpgtPath = tpgtPath
+		istats.Tpgt[ii].IsEnable, _ = isPathEnable(tpgtPath)
+		if istats.Tpgt[ii].IsEnable {
+			matchesLunsPath, _ := getLun(tpgtPath)
+			istats.Tpgt[ii].Luns = make([]LUN, len(matchesLunsPath))
+
+			for ll, lunPath := range matchesLunsPath {
+				backstore, objectName, typeNumber, err := getLunLinkTarget(lunPath)
```

Comment:
I think it would be better here to return a `(LUN, err)` from `getLunLinkTarget()` and then do something like `istats.Tpgt[ii].Luns[ll] = lun`

Reply:
Do you mean, I should create another function to call on top? Because I still need do this dynamic assignment inside that function to return the LUN list and passing the Tgpt object inside that function to return LUN object into it.

Reply:
What I mean is just change the return type of `getLunLinkTarget()` to something like `(*LUN, error)`, so instead of receiving multiple return values and then assigning them individually.  You can just assign the whole struct to the array.  And if there is an error, you can return `nil` for the lun struct.

Reply:
The contents of the inner loop becomes something like this:
```golang
	for j, lunPath := range matchesLunsPath {
		lun, err := getLunLinkTarget(lunPath)
		if err != nil {
			continue
		}
		istats.Tpgt[i].Luns[j] = *lun
	}
```

Reply:
Or could maybe simplify a bit?
```golang
	for _, lunPath := range matchesLunsPath {
		lun, err := getLunLinkTarget(lunPath)
		if err == nil {
			istats.Tpgt[i].Luns = append(istats.Tpgt[i].Luns, lun)
		}
	}
```

Reply:
Lovely I can do this for sure. I will update it.